conformance-test-framework/src/main/java/org/jspecify/conformance/ExpectedFact.java

```diff
@@ -0,0 +1,167 @@
+// Copyright 2023 The JSpecify Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.jspecify.conformance;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.stream.Collectors.joining;
+
+import com.google.common.collect.ImmutableList;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * An assertion that the tool behaves in a way consistent with a specific fact about a line in the
+ * source code. Some of these facts indicate that according to the JSpecify specification, the code
+ * in question may have an error that should be reported to users; other expected facts are
+ * informational, such as the expected nullness-augmented type of an expression.
+ */
+public final class ExpectedFact extends Fact {
+
+  private static final Pattern NULLNESS_MISMATCH =
+      Pattern.compile("jspecify_nullness_mismatch\\b.*");
+
+  private static final ImmutableList<Pattern> ASSERTION_PATTERNS =
+      ImmutableList.of(
+          NULLNESS_MISMATCH,
+          // TODO: wildcard types have whitespace
+          Pattern.compile("test:cannot-convert:\\S+ to \\S+"),
+          Pattern.compile("test:expression-type:[^:]+:.*"),
+          Pattern.compile("test:irrelevant-annotation:\\S+"),
+          Pattern.compile("test:sink-type:[^:]+:.*"));
+
+  private final String fact;
+  private final long factLineNumber;
+
+  ExpectedFact(Path file, long lineNumber, String fact, long factLineNumber) {
+    super(file, lineNumber);
+    this.fact = fact;
+    this.factLineNumber = factLineNumber;
+  }
+
+  /**
+   * Returns an expected fact representing that the source type cannot be converted to the sink type
```

Comment:
From this description I expected the return type to be `ExpectedFact`. Is there a better term for the returned String?

Reply:
Yes. Based on this I've tried to use "factText" throughout, and merged `ReportedFact.expectedFact()` into `ReportedFact.getFactText()`.

I also noticed that these static methods are only used from implementations of `ReportedFact`, so I moved them there.